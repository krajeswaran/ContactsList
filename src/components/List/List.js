import React, {useState, useEffect, useRef} from 'react';
import {
    SafeAreaView,
    Text,
    FlatList,
    View,
    StyleSheet,
    Image,
    TouchableOpacity,
    Modal
} from 'react-native';
import {Picker} from '@react-native-picker/picker';
import { SearchBar } from 'react-native-elements';
import {matchSorter} from 'match-sorter';


//Component for each contact that gets rendered
const Contact = ({
    title,
    phone_numbers,
    userpic,
    navigation
}) => {
    const [modalVisible,setModalVisible] = useState(false);
    const [selected_contact,setSelectedContact] = useState(null);

    const [pickerval,setPickerVal] = useState(0);
    const [multiple,setMultiple] = useState(false);

    //First selection of a contact
    const selectContact = (title,phone_numbers) => {
        //If there are no phone numbers present then the contact is not processed further
        if(phone_numbers.length > 0){
            const selected_obj = {title:title,phone_numbers:phone_numbers}
            setSelectedContact({...selected_obj});
            if(phone_numbers.length > 1){
                setMultiple(true);
            }
            setModalVisible(true);
        }
        else{
            alert("This contact does not have any phone number")
        }
    }

    //Confirming the selection
    const confirmSelection = () => {
        setModalVisible(false);
        const selected_obj = {title:selected_contact.title,phone_number:selected_contact.phone_numbers[pickerval]}
        navigation.navigate('Home',{confirmed_select:selected_obj})
    }

    return(
        <>
        {/* Modal view for confirming the selection and selecting from multiple numbers if any */}
        <Modal
            animationType="slide"
            transparent={true}
            visible={modalVisible}
            onRequestClose={() => {
                setModalVisible(false)
            }}
        >
            <View style={styles.centeredView}>
            <View style={styles.modalView}>
                <Text style={styles.modalText}>{selected_contact ? selected_contact.title : ''}</Text>

                {selected_contact ?
                    <View style={styles.selected_contact_numbers}>

                        {/*If the phone numbers are multiple, a picker option is opened below for selection 
                        if not then, the user can directly confirm their selection of the contact */}
                        {phone_numbers.map((phone_no,index) =>(
                            <View key={index} style={styles.phone_container}>
                                <Text style={styles.selected_contact_title}>
                                    {phone_no.label ? phone_no.label[0].toUpperCase() + phone_no.label.slice(1) : ''}
                                </Text>
                                <Text style={styles.selected_contact_content}>
                                    {phone_no.number}
                                </Text>
                            </View>
                        ))}
                        {multiple && <Text style={styles.multipleAlert}>Note: Contact has multiple numbers, select one before confirming selection</Text>}
                        {multiple && <Picker
                            selectedValue={pickerval}
                            style={styles.pickerStyle}
                            onValueChange={(itemValue, itemIndex) =>
                                setPickerVal(itemIndex)
                            }>
                            {phone_numbers.map((phone_no,index) =>(
                                <Picker.Item key={index} label={phone_no.number} value={index} />
                            ))}
                        </Picker>}
                    </View> :
                    <Text>
                    </Text>
                }
                {/* Confirm Selection and Close Modal Buttons */}
                <TouchableOpacity
                    style={{ ...styles.openButton, backgroundColor: "#2196F3", marginTop:30 }}
                    onPress={() => {
                        confirmSelection();
                    }}
                    >
                    <Text style={styles.textStyle}>Confirm Selection</Text>
                </TouchableOpacity>
                <TouchableOpacity
                style={{ ...styles.openButton, backgroundColor: "#2196F3", marginTop:20 }}
                onPress={() => {
                    setModalVisible(!modalVisible);
                    setMultiple(false);
                }}
                >
                <Text style={styles.textStyle}>Cancel</Text>
                </TouchableOpacity>
            </View>
            </View>
        </Modal>

        {/* Singular contact view */}
        <TouchableOpacity onPress={()=>selectContact(title,phone_numbers)}>
            <View style={styles.contact_item}>
                <View style={styles.contact_container}>
                    <View style={styles.image_container}>
                        <Image style={styles.contact_image} source={{uri:userpic}} />
                    </View>
                    <Text style={styles.contact_title}>
                        {title}
                    </Text>
                </View>
                <View style={styles.top_spacing}>
                    {phone_numbers.map((phone_no,index)=>(
                        <View key={index} style={styles.phone_container}>
                            <Text style={styles.selected_contacted_title}>
                                {phone_no.label ? phone_no.label[0].toUpperCase() + phone_no.label.slice(1) : ''}
                            </Text>
                            <Text style={styles.selected_contact_content}>
                                {phone_no.number}
                            </Text>
                        </View>
                    ))}
                </View>
            </View>
        </TouchableOpacity>
        </>
    )
}

const List = ({contacts,navigation}) => {

    const [searchval,setSearchVal] = useState('');
    const [filtered,setFiltered] = useState([]);

    //rendering each item from flatlist
    const renderItem = ({item}) => (
        <Contact 
            title={item.fullName} 
            phone_numbers={item.phoneNumbers} 
            userpic={item.hasThumbnail ? item.thumbnailPath : 'https://thepeakid.com/wp-content/uploads/2016/03/default-profile-picture-300x300.jpg'}
            navigation={navigation}
        />
      );

    const updateSearchText = (val) => {
        setSearchVal(val)
    }

    useEffect(() => {
        if(searchval.length > 0){
            //This is the search functionality, where the input length if available, the array is sorted/searched through using the
            //fullName key, and results are displayed accordingly
            flatListRef.current.scrollToOffset({ animated: true, offset: 0 })
            setFiltered(matchSorter(contacts, searchval, {keys: ['fullName']}))
        }
    }, [searchval])


    const flatListRef = useRef();

    //ref to flatlist has been declared to resume its position from the top when something gets searched

    //Optimisation for flat list has been used with getItemLayout, to give a fixed height to each item when listing
    //Helps in cases of hundreds of rows of data

    return(
        <SafeAreaView style={{flex:1}}>
            <View style={styles.container}>
                <SearchBar
                    placeholder="Search by Name ..."
                    onChangeText={updateSearchText}
                    value={searchval}
                    lightTheme={true}
                />
                {contacts.length > 0 ? 
                    <FlatList
                        ref={flatListRef}
                        keyboardShouldPersistTaps='always'
                        data={filtered.length > 0 ? filtered : contacts}
                        renderItem={renderItem}
                        keyExtractor={item => item.recordID}
                        initialNumToRender={5}
                        style={styles.flat_list}
                        getItemLayout={(data, index) => (
                            {length: 150, offset:150 * index, index}
                        )}
                    /> :
                    <Text style={{marginTop:20, textAlign:'center'}}>
                        No contacts are present
                    </Text>
                }
            </View>
        </SafeAreaView>
    )
}

const styles = StyleSheet.create({
   contact_item:{
       backgroundColor:'white',
       marginTop:5,
       marginBottom:10,
       marginLeft:10,
       marginRight:10,
       borderColor:'lightgrey',
       borderWidth:1,
       borderStyle:'solid',
       padding:12,
       height:150
   },
   phone_container:{
       display:'flex',
       flexDirection:'row',
       marginTop:2
   },
   contact_image:{
       height:50,
       resizeMode:'contain'
   },
   container:{
       backgroundColor:'#f8f8f8',
       marginBottom:20,
       flex:1
   },
   centeredView: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    marginTop: 22
  },
  modalView: {
    margin: 20,
    backgroundColor: "white",
    borderRadius: 20,
    padding: 35,
    alignItems: "center",
    shadowColor: "#000",
    shadowOffset: {
      width: 0,
      height: 2
    },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
    elevation: 5
  },
  openButton: {
    backgroundColor: "#F194FF",
    borderRadius: 20,
    padding: 10,
    elevation: 2
  },
  textStyle: {
    color: "white",
    fontWeight: "bold",
    textAlign: "center"
  },
  modalText: {
    marginBottom: 15,
    textAlign: "center",
    fontWeight:'bold'
  },
  selected_contact_numbers:{
      justifyContent:'center',
      alignItems:'center'
  },
  selected_contacted_title:{
      fontWeight:'bold',
      fontSize:13
  },
  selected_contact_content:{
      marginLeft:5,
      fontSize:13
  },
  multipleAlert:{
      fontSize:11,
      color:'red',
      marginTop:30,
      textAlign:'center',
      fontWeight:'bold'
  },
  contact_container:{
      display:'flex',
      justifyContent:'flex-start',
      flexDirection:'row',
      alignItems:'center'
  },
  contact_title:{
      fontSize:16,
      fontWeight:'bold',
      marginLeft:12
  },
  top_spacing:{
      marginTop:10
  },
  image_container:{
    height:50,
    width:50
  },
  pickerStyle:{
      width:200
  },
  flat_list:{
      marginBottom:20
  }
})

export default List;