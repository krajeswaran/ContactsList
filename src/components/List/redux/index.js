import {Creators} from './actions';
import Contacts from 'react-native-contacts';
import {matchSorter} from 'match-sorter';
import { PermissionsAndroid, Platform } from 'react-native';
import {check, request, PERMISSIONS, RESULTS} from 'react-native-permissions';

const all_contacts = Creators.all_contacts;

const get_all_contacts = () => {

	return dispatch => {

        //Async action performed to get all contacts and update app store
        //Permissions behave differently for Android and iOS hence declared separately
        //Action is being dispatched with all contacts results

        const get_contact_permissions_android = () => {

          check(PERMISSIONS.ANDROID.READ_CONTACTS)
    .then((result) => {
      switch (result) {
        case RESULTS.UNAVAILABLE:
          console.log(
            'This feature is not available (on this device / in this context)',
          );
          break;
        case RESULTS.DENIED:
          console.log(
            'The permission has not been requested / is denied but requestable',
          );
      request(
                    PERMISSIONS.ANDROID.READ_CONTACTS
                ).then(req_result=>{
                    if(req_result === 'granted'){
                        Contacts.getAll().then(contacts => {
            
                            //Sorting with full name, is what will be used for display order and search
                            let _contacts = contacts.map(obj=> ({ ...obj, fullName: obj.givenName + " " + obj.familyName })); 
                
                            dispatch(all_contacts(matchSorter(_contacts, '', {keys: ['fullName']})));  
                          })                           
                    }
                    else{
                        alert("The app requires your permission to read contacts")
                    }
                });
          break;
        case RESULTS.GRANTED:
          console.log('The permission is granted');
          break;
        case RESULTS.BLOCKED:
          console.log('The permission is denied and not requestable anymore');
          break;
      }
    })
    .catch((error) => {
      console.log("Permissions error",error)
    });              
    }

    if(Platform.OS === 'ios'){
        Contacts.getAll().then(contacts => {
            // contacts returned

            //Contacts are being mapped through to add another full name field
            let _contacts = contacts.map(obj=> ({ ...obj, fullName: obj.givenName + " " + obj.familyName })); 

            //Sorting based on the name of the full field
            dispath(all_contacts(matchSorter(_contacts, '', {keys: ['fullName']}))); 
          })   
      }
      else{
        //Permissions from Android to read contacts
        try{
            get_contact_permissions_android();
        }
        catch(e){
            console.log(e)
        }
      }
	}

};

export default { get_all_contacts }