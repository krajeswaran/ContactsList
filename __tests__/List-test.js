import React from 'react';
import {ActivityIndicator, FlatList, Text, TextInput} from 'react-native';
import renderer from 'react-test-renderer';
import List from '../src/components/List/List';

jest.setTimeout(15000);

it('renders correctly', () => {
  const tree = renderer.create(<List />).toJSON();
  expect(tree).toMatchSnapshot();
});