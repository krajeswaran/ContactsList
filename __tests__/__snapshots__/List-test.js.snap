// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#f8f8f8",
        "marginBottom": 20,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#e1e8ee",
          "borderBottomColor": "#e1e1e1",
          "borderBottomWidth": 1,
          "borderTopColor": "#e1e1e1",
          "borderTopWidth": 1,
          "padding": 8,
        }
      }
    >
      <View
        style={
          Object {
            "paddingHorizontal": 0,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#bdc6cf",
              "borderBottomWidth": 0,
              "borderColor": "#86939e",
              "borderRadius": 3,
              "flexDirection": "row",
              "minHeight": 30,
              "overflow": "hidden",
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "height": 40,
                "justifyContent": "center",
                "marginLeft": 8,
                "marginVertical": 4,
                "paddingRight": 4,
              }
            }
          >
            <View
              style={
                Object {
                  "overflow": "hidden",
                }
              }
            >
              <View>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Text
                    allowFontScaling={false}
                    style={
                      Array [
                        Object {
                          "color": "#86939e",
                          "fontSize": 18,
                        },
                        Object {
                          "backgroundColor": "transparent",
                        },
                        Object {
                          "fontFamily": "Material Icons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        Object {},
                      ]
                    }
                    testID="iconIcon"
                  >
                    î¢¶
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <TextInput
            allowFontScaling={true}
            editable={true}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="Search by Name ..."
            placeholderTextColor="#86939e"
            platform="default"
            rejectResponderTermination={true}
            style={
              Object {
                "alignSelf": "center",
                "color": "#86939e",
                "flex": 1,
                "fontSize": 18,
                "marginLeft": 10,
                "minHeight": 40,
              }
            }
            testID="searchInput"
            underlineColorAndroid="transparent"
            value=""
          />
          <View
            style={
              Object {
                "alignItems": "center",
                "height": 40,
                "justifyContent": "center",
                "marginRight": 8,
                "marginVertical": 4,
                "paddingRight": 4,
              }
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                }
              }
            />
          </View>
        </View>
        <Text
          style={
            Object {
              "color": "#ff190c",
              "fontSize": 12,
              "height": 0,
              "margin": 0,
              "padding": 0,
            }
          }
        />
      </View>
    </View>
    <RCTScrollView
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      getItemLayout={[Function]}
      horizontal={false}
      initialNumToRender={5}
      keyExtractor={[Function]}
      keyboardShouldPersistTaps="always"
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "marginBottom": 20,
        }
      }
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View />
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
